lib/storage/src/rbac/mod.rs

```diff
@@ -1,9 +1,209 @@
+
+    pub access: CollectionAccessMode,
+
+    /// Payload constraints.
+    /// An object where each key is a JSON path, and each value is JSON value.
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub payload: Option<PayloadConstraint>,
+}
+
+#[derive(Serialize, Deserialize, Eq, PartialEq, Ord, PartialOrd, Copy, Clone, Debug)]
+pub enum ClusterAccessMode {
+    /// Read-only access to the whole cluster.
+    #[serde(rename = "r")]
+    Read,
+
+    /// Read and write access to the whole cluster.
+    #[serde(rename = "rw")]
+    ReadWrite,
+}
+
+#[derive(Serialize, Deserialize, Eq, PartialEq, Ord, PartialOrd, Copy, Clone, Debug)]
+pub enum CollectionAccessMode {
+    /// Read-only access to a collection.
+    #[serde(rename = "r")]
+    Read,
+
+    /// Read and write access to a collection, with some restrictions.
+    #[serde(rename = "rw")]
+    ReadWrite,
+
+    /// Read and write access to a collection, with no restrictions.
+    #[serde(rename = "m")]
+    Manage,
+}
+
+#[derive(Serialize, Deserialize, PartialEq, Clone, Debug)]
+pub struct PayloadConstraint(pub HashMap<JsonPath, ValueVariants>);
+
+impl Access {
+    /// Create an `Access` object with full access.
+    /// The ``_reason`` parameter is not used in the code, but serves as a mandatory commentary to
+    /// explain why the access is granted, e.g. ``Access::full("Internal API")`` or
+    /// ``Access::full("Test")``.
+    pub const fn full(_reason: &'static str) -> Self {
+        Self::Cluster(ClusterAccessMode::ReadWrite)
+    }
+
+    pub const fn full_ro(_reason: &'static str) -> Self {
+        Self::Cluster(ClusterAccessMode::Read)
+    }
+
+    /// Check if the user has access to the whole cluster.
+    pub fn check_cluster_access(
+        &self,
+        min_mode: ClusterAccessMode,
+    ) -> Result<CollectionMultipass, StorageError> {
+        match self {
+            Access::Cluster(mode) if *mode >= min_mode => Ok(CollectionMultipass),
+            _ => Err(StorageError::unauthorized("Cluster access is not allowed")),
+        }
+    }
+
+    /// Check if the user has access to a collection with given constraints.
+    ///
+    /// If ``whole_collection`` is true, then the access should not limited by a payload
+    /// restriction.
+    pub fn check_collection_access<'a>(
+        &self,
+        collection_name: &'a str,
+        whole_collection: bool,
+        mode: CollectionAccessMode,
+    ) -> Result<CollectionPass<'a>, StorageError> {
+        match self {
+            Access::Cluster(ClusterAccessMode::Read) => {
+                if mode != CollectionAccessMode::Read {
+                    return Err(StorageError::unauthorized(
+                        "Only read-only access is allowed",
+                    ));
+                }
+            }
+            Access::Cluster(ClusterAccessMode::ReadWrite) => (),
+            Access::Collection(rules) => {
+                let view = rules.find_view(collection_name)?;
```

Comment:
This search is not exhaustive, as there can be more than 1 rule for the same collection

Reply:
Could you provide a use-case/reasoning for that?

I was thinking in a different way: rules should be additive, meaning that adding a new rule to the list should not impose new restrictions but rather lift them.

As for multiple rules with the same collection, we could use any rule of the matching type, e.g. use the first one for read-only APIs, and the second for writes:
```json
"access": [
  {"collections":["a"], "access": "r"},
  {"collections":["a"], "access": "rw", "payload": {"x": 0}}
]
```
But it would be complicated, so in this PR I decided to ignore this and just always use the first rule.

Reply:
> I was thinking in a different way: rules should be additive, meaning that adding a new rule to the list should not impose new restrictions but rather lift them.

This is also possible, but after some thought, it is not trivial to combine 

My reasoning is, for this example: 
```json
"access": [
  {"collections":["a", "b"], "access": "rw"},
  {"collections":["a", "c"], "access": "r", "payload": {"x": 0}}
]
```

The view for collection "a" generated for this can be:
```json
// Option 1: additive (permissive)
{ "collection": "a", "access": "rw" }

// Option 2: subtractive (restrictive)
{ "collection": "a", "access": "r", "payload": { "x": 0 } }

// Option 3: hierarchical (global access > collection access > payload constraints)
{ "collection": "a", "access": "rw", "payload": { "x": 0 } }
```

Now, if we also have a third config like `{"collections":["a"], "access": "rw", "payload": {"y": 0}}`. How would payload combination work?

I believe the easiest solution is to validate the access struct itself to make sure there aren't multiple configs for the same collection, wdyt?

Reply:
> I believe the easiest solution is to validate the access struct itself to make sure there aren't multiple configs for the same collection, wdyt?

LGTM